### ISSUE TITLE
Changing KLA

### ISSUE THREAD

**Issue by HughP** [HughP/dnj-corpus#39]
@iandoug is there a way to put your own keyboards into KLA? If so how?

**Comment by iandoug** [comment]
Yes, you need to edit the keys ... left-click the key and dialogue should pop up letting you edit what's on each slot on the key. Alternatively you can import an existing KLA json file. Just remember to save / export when done because the changes will only exist in memory. 

If you mean "physical" layouts, that's more of a problem, since each physical layout is hard-coded. By default KLA supports ANSI/ISO and Ergodox/Kinesis hardware form factors, Den and I added Ergolinear and Matrix as well. We have been talking about a more generic solution (ideally with full interchange between KLA and KLE) but nothing available yet in that regard.

I assume you're aiming at a Macbook layout?

**Comment by HughP** [comment]
But there doesn't seem to be a way to take the personalized keyboard option and save it's json layout.

**Comment by iandoug** [comment]
Click "export", which pops up a window with the json, click select all, then ctrl-c / copy, and paste into a text editor on your laptop. Then save. Reverse the process to reload it later...
If you are working on Patrick's original KLA then you can use the "share results" button, which saves everything in the cloud for later access.

**Comment by HughP** [comment]
Also somewhere You pointed me to an "improved" version of KLA. What were those improvements and how did they work?

**Comment by iandoug** [comment]
Re saving the "personalised" layout, don't think that is possible, it's usually not any good anyway (well, in the testing we've done, we've always gotten better layouts ourselves). So you would need to print the screenshot and then load it manually as a custom layout.

The improved versions are Den's versions, currently three:
http://shenafu.com/code/keyboard/Keyboard%20Layout%20Analyzer.html#/main
http://shenafu.com/code/keyboard/Keyboard%20Layout%20Analyzer%202.html#/main
http://shenafu.com/code/keyboard/kla3/
and the "test" version:
http://shenafu.com/code/keyboard/klatest/#/main

V1 added counting vertical distance travelled (assumes Cherry-style switches, now Macbook scissor switches)
V2 I didn't like, he started changing parameters, punishing pinkies and index fingers more
V3 similar to V2, but I managed to get good layout on that.
Test: more radical ideas in terms of the scoring. Not popular at the moment :-)

I don't think Patrick's Personalized algos will cope with the peculiarities of Dan and Mephaa.
